chainer/utils/conv_nd_kernel.py

```diff
@@ -0,0 +1,299 @@
+        #     int out_x1_0 = min(out_0, (x_0       + s_0) / s_0);
+        #     int out_x0_1 = max(0,     (x_1 - k_1 + s_1) / s_1);
+        #     int out_x1_1 = min(out_1, (x_1       + s_1) / s_1);
+        #     ... Before-part here ...
+        #     for (int out_x_0 = out_x0_0; out_x_0 < out_x1_0; ++out_x_0) {
+        #       int kx_0 = x_0 - out_x_0 * s_0 + k_0 * c0;
+        #       for (int out_x_1 = out_x0_1; out_x_1 < out_x1_1; ++out_x_1) {
+        #         int kx_1 = x_1 - out_x_1 * s_1 + k_1 * kx_0;
+        #         ... Main-part here ...
+        #       }
+        #     }
+        #     ... After-part here ...
+        def aux(out_x0, out_x1, out, x, k, s):
+            return [
+                'int {} = max(0,     ({} - {} + {}) / {});'.format(
+                    out_x0, x, k, s, s),
+                'int {} = min({}, ({}       + {}) / {});'.format(
+                    out_x1, out, x, s, s)]
+        out_x0s = _vars('out_x0', self.ndim)
+        out_x1s = _vars('out_x1', self.ndim)
+        bounds = sum(map(
+            aux, out_x0s, out_x1s, self.outs, xs, self.ks, self.ss), [])
+
+        def _loop_main(main):
+            w = _writer()
+
+            # Loop openings.
+            out_xs = _vars('out_x', self.ndim)
+            kxs = _vars('kx', self.ndim)
+            kxs1 = ['c0'] + kxs[:-1]
+            for (out_x, out_x0, out_x1, kx, s, x, k, kx1) in zip(
+                    out_xs, out_x0s, out_x1s, kxs, self.ss, xs, self.ks, kxs1):
+                w('for (int {} = {}; {} < {}; ++{}) {{'.format(
+                    out_x, out_x0, out_x, out_x1, out_x), 'inc')
+                w('int {} = {} - {} * {} + {} * {};'.format(
+                    kx, x, out_x, s, k, kx1))
+
+            # Main-part.
+            kx = kxs[-1]
+            for l in main(kx, out_xs).split('\n'):
+                w(l)
+
+            # Loop closings.
+            for _ in out_xs:
+                w('}', 'dec')
+
+            return [w()]
+
+        return bounds, _loop_main
+
+    def _compile_procedure(self, xs):
+        # 2D: val = val + col[(out_x_1 + out_1 * (out_x_0 + out_0 * kx_1))];
+        def _main(kx, out_xs):
+            index = muladdexp(self.outs, out_xs, kx)
+            return 'val = val + col[{}];'.format(index)
+        before = ['T val = 0;']
+        after = ['img = val;']
+        return before, _main, after
+
+    def _operation(self):
+        c0 = self._compile_c0()
+        x, xs = self._compile_x()
+        loop_bounds, loop_main = self._compile_loop(xs)
+        before, main, after = self._compile_procedure(xs)
+        return '\n'.join(
+            c0 + x + loop_bounds + before + loop_main(main) + after)
+
+    def generate(self):
+        in_params = self._in_params()
+        out_params = self._out_params()
+        operation = self._operation()
+        name = 'col2im_{}d'.format(self.ndim)
+        return in_params, out_params, operation, name
+
+
+# just for debug.
+if __name__ == "__main__":
```

Comment:
I left it for Chainer developers' convenience, I will remove it because it is now simple enough, it was more complicated during implementing this module.